src/main/java/io/debezium/connector/jdbc/JdbcChangeEventSink.java

```diff
@@ -0,0 +1,406 @@
+            transaction.rollback();
+            throw e;
+        }
+    }
+
+    private int bindFieldValue(NativeQuery<?> query, FieldDescriptor field, Object value, int index) {
+        return index + dialect.bindValue(field, query, index, value);
+    }
+
+    private int bindKeyValuesToQuery(SinkRecordDescriptor record, NativeQuery<?> query, int index) {
+        if (PrimaryKeyMode.KAFKA.equals(config.getPrimaryKeyMode())) {
+            query.setParameter(index++, record.getTopicName());
+            query.setParameter(index++, record.getRawRecord().kafkaPartition());
+            query.setParameter(index++, record.getRawRecord().kafkaOffset());
+        }
+        else if (!record.getKeyFieldNames().isEmpty()) {
+            // There are key fields, either mapped to RECORD_KEY or RECORD_VALUE potentially
+            if (PrimaryKeyMode.RECORD_KEY.equals(config.getPrimaryKeyMode())) {
+                final Schema keySchema = record.getRawRecord().keySchema();
+                if (keySchema != null && Schema.Type.STRUCT.equals(keySchema.type())) {
+                    for (String fieldName : record.getKeyFieldNames()) {
+                        final FieldDescriptor field = record.getFields().get(fieldName);
+                        final Object fieldValue = ((Struct) record.getRawRecord().key()).get(fieldName);
+                        index = bindFieldValue(query, field, fieldValue, index);
+                    }
+                }
+                else {
+                    throw new ConnectException("Cannot bind primary key with no struct-based key");
+                }
+            }
+            else if (PrimaryKeyMode.RECORD_VALUE.equals(config.getPrimaryKeyMode())) {
+                final Schema valueSchema = record.getRawRecord().valueSchema();
+                if (valueSchema != null && Schema.Type.STRUCT.equals(valueSchema.type())) {
+                    final Struct source;
+                    if (record.isDebeziumSinkRecord()) {
+                        source = ((Struct) record.getRawRecord().value()).getStruct(Envelope.FieldName.AFTER);
+                    }
+                    else {
+                        source = ((Struct) record.getRawRecord().value());
+                    }
+
+                    for (String fieldName : record.getKeyFieldNames()) {
+                        final FieldDescriptor field = record.getFields().get(fieldName);
+                        final Object fieldValue = source.get(fieldName);
+                        index = bindFieldValue(query, field, fieldValue, index);
+                    }
+                }
+                else {
+                    throw new ConnectException("Cannot bind primary key with a non-struct value");
+                }
+            }
+        }
+        return index;
+    }
+
+    private int bindNonKeyValuesToQuery(SinkRecordDescriptor record, NativeQuery<?> query, int index) {
+        final Struct struct;
+        if (record.isDebeziumSinkRecord()) {
```

Comment:
`SinkRecordDescriptor` should have a method that will return the correct `Struct` based on Debezium/non-Debezium payload